### PR TITLE
LibJS: Minor operator improvements

### PR DESCRIPTION
**Pull request by 0xtechnobabble**
Just a small and simple patch that adds a few operators and moves the logical `not` operator to a new `UnaryExpression` class.

### CODE REVIEW

File: Libraries/LibJS/AST.cpp [hunk 1]
```diff
@@ -120,6 +172,24 @@ Value BinaryExpression::execute(Interpreter& interpreter) const
         return sub(lhs_result, rhs_result);
     case BinaryOp::TypedEquals:
         return typed_eq(lhs_result, rhs_result);
+    case BinaryOp::TypedInequals:
+        return Value(!typed_eq(lhs_result, rhs_result).as_bool());
+    case BinaryOp::Greater:
+        return greater(lhs_result, rhs_result);
+    case BinaryOp::Smaller:
+        return Value(!greater(lhs_result, rhs_result).as_bool());
```
Comment (sunverwerth) [inline Libraries/LibJS/AST.cpp]: Careful here. The negation of `>` is `<=` not `<`.

File: Libraries/LibJS/AST.cpp [hunk 2]
```diff
@@ -177,6 +247,38 @@ void BinaryExpression::dump(int indent) const
...
+        op_string = "^";
+        break;
+    case BinaryOp::BitLeftShift:
+        op_string = "<<";
+        break;
+    case BinaryOp::BitRightShift:
+        op_string = ">>";
+        break;
+    case BinaryOp::BitNot:
+        op_string = "~";
```
Comment (sunverwerth) [inline Libraries/LibJS/AST.cpp]: BitNot is a unary operator and should have its own class.

File: Libraries/LibJS/AST.cpp (hunk 1 above)
Comment (0xtechnobabble) [reply Libraries/LibJS/AST.cpp]: Oops, that's an embarrassing error on my end :sweat_smile: 

Anyways, fixed!

File: Libraries/LibJS/AST.cpp (hunk 2 above)
Comment (0xtechnobabble) [reply Libraries/LibJS/AST.cpp]: You're definitely right, done, I'll also take the liberty of moving the `not` logical operator to the new unary class.

**Comment by 0xtechnobabble** [comment]
Alright, everything should be fixed now, sorry for the mess :P

**Comment by sunverwerth** [comment]
Nah, don't be sorry. Every contribution counts :)

Any reason you introduced a seperate `LogicalExpression` class? I feel like this will just be a copy of `BinaryExpression`.

**Comment by 0xtechnobabble** [comment]
Well, since Andreas used the esprima AST design as a guide, I just did the same :P, but I also think separating the `LogicalExpression` class is reasonable since logically speaking, binary expressions are supposed to return numbers (normally) while as the logical ones are supposed to return a boolean.

**Comment by sunverwerth** [comment]
It's fine, I was just worrying about duplicated code in the future. Is esprima based on the official ECMA spec?

**Comment by 0xtechnobabble** [comment]
Well, there is no official ECMAScript AST spec that I'm aware of, but there is a community one at [estree/estree](https://github.com/estree/estree) that also specifies the [separation of logical expressions.](https://github.com/estree/estree/blob/master/es5.md#logicalexpression). If you were talking about the esprima parser's support, the project officially provides full support for ECMAScript 2016.

**Comment by awesomekling** [comment]
Looks good, thank you :) :+1:

**Pull request merged**